src/lib/iconv/Converter.cxx

```diff
@@ -0,0 +1,242 @@
+				return false;
+			}
+		} else if ((c & three_byte_mask) == three_byte) {
+			if (!(is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++]))) {
+				return false;
+			}
+		} else if ((c & four_byte_mask) == four_byte) {
+			if (!(is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++]))) {
+				return false;
+			}
+		} else if ((c & five_byte_mask) == five_byte) {
+			if (!(is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++]))) {
+				return false;
+			}
+		} else if ((c & six_byte_mask) == six_byte) {
+			if (!(is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++])
+				&& is_utf8_special_byte(str[i++]))) {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool
+is_gb2312_code(const unsigned char *str, unsigned size)
+{
+	unsigned char c;
+
+	for (unsigned i=0; i<size;) {
+		c = str[i++];
+		if (is_ascii(c)) {
+			continue;
+		} else if (c >= 0xA1 && c <= 0xF7) {
+			c = str[i++];
+			if (!(c >= 0XA1 && c <= 0XFE)) {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool
+is_big5_code(const unsigned char *str, unsigned size)
+{
+	unsigned char c;
+
+	for (unsigned i=0; i<size;) {
+		c = str[i++];
+		if (is_ascii(c)) {
+			continue;
+		} else if (c >= 0xA1 && c <= 0xF9) {
+			c = str[i++];
+			if (!((c >= 0x40 && c <= 0x7E)
+				|| (c >= 0XA1 && c <= 0xFE))) {
+				return false;
+			}
+		} else {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+static bool
+is_gbk_code(const unsigned char *str, unsigned size)
+{
+	unsigned char c;
+
+	for (unsigned i=0; i<size;) {
+		c = str[i++];
+		if (is_ascii(c)) {
+			continue;
+		} else if (c >= 0x81 && c <= 0xFE) {
+			c = str[i++];
+			if (!(c >= 0x40 && c <= 0xFE)) {
+				return false;;
+			}
+		} else {
+			return false;
+		}
+	}
+
+	return true;
+}
+
+std::string
+Converter::toUTF8(const char *src, size_t inlen)
+{
+	if (src == nullptr ||
+		inlen == 0) {
+		return std::string();
+	}
+
+	Code c;
+	const unsigned char *buf = (const unsigned char *)src;
+
+	if (is_utf8_code(buf, inlen)) {
+		return std::string(src, inlen);
+	} else if (is_gb2312_code(buf, inlen)) {
+		c = GB2312;
+	} else if (is_big5_code(buf, inlen)) {
+		c = BIG5;
+	} else if (is_gbk_code(buf, inlen)) {
+		c = GBK;
+	} else {
+		FormatError(domain, "unkown code: %s", src);
+		return std::string(src, inlen);
+	}
+	if (handle != nullptr && c != code) {
+		iconv_close(handle);
+		handle = nullptr;
+	}
+	code = c;
+	if (handle == nullptr) {
+		handle = iconv_open(code_str[UTF8], code_str[code]);
+		if (handle == nullptr) {
+			FormatError(domain, "fail iconv_open, code:%s", code_str[code]);
+			return std::string(src, inlen);
+		}
+	}
+
+	char dst[4096];
+	memset(dst, 0, sizeof(dst));
```

Comment:
Why?

Reply:
Just for caryaudio.  Convert some chinese chaset, which may GB2312, BIG5 or GBK coded,  to UTF-8.